csrc/kernel_cache.cpp

```diff
@@ -746,20 +745,91 @@ std::vector<at::Tensor> FusionKernelRuntime::runWithInputs(
   return fusion_outputs;
 }
 
-std::unordered_map<Val*, const ArgAbstract*> FusionKernelRuntime::
-    runSegmentsWithInputs(KernelArgumentHolder& args, bool is_dry_run) {
-  TORCH_INTERNAL_ASSERT(
-      args.size() == segmented_fusion_->inputs().size(),
-      "Inputs were not set up correctly, received ",
-      args.size(),
-      " inputs but expected ",
-      segmented_fusion_->inputs().size());
-
-  std::unordered_map<Val*, const ArgAbstract*> tensor_map =
-      mapFusionInputsToArgs(args);
+namespace {
+// This ArgumentManager do two things
+// (1) add outputs from a segment to the global fusion args to pass it to next
+// segment (2) delete args no longer being used to save memory. For task (2), it
+// checks the input and output arguments of each segment and make a map from val
+// to the segment_id where the val is lastly used. The arguments representing
+// these vals are then deleted after the segment runs.
+class ArgumentManager {
+ public:
+  ArgumentManager(
+      KernelArgumentHolder& args,
+      const std::vector<SegmentedGroup*>& group_run_order,
+      std::unordered_map<Val*, const ArgAbstract*>& tensor_map)
+      : fusion_args_(args),
+        group_run_order_(group_run_order),
+        tensor_map_(tensor_map) {
+    setLastUsedSegmentID();
+  }
+  template <typename T>
```

Comment:
Please comment that `T` is assumed to be either `std::vector<at::Tensro>` or `KernelArgumentHolder`. We should probably try to make the output type uniform no matter it's a real or dry run, but that's a different topic.

Reply:
added